File: DataStructures/java/doublyLinkedList/DoublyLinkedList.java

public class DoublyLinkedList {

    private int length;
    private DoublyNode first;
    private DoublyNode last;

    public DoublyLinkedList() {
        this.length = 0;
        this.first = null;
        this.last = null;
    }

    public void insertToFirst(int item) {
        DoublyNode newNode = new DoublyNode();
        newNode.item = item;
        if (isEmpty()) {
            first = last = newNode;
            first.prev = null;
            first.next = null;
        } else {
            newNode.next = first;
            first.prev = newNode;
            first = newNode;
            newNode.prev = null;
        }
        length++;
    }

    public void insertToLast(int item) {
        DoublyNode newNode = new DoublyNode();
        newNode.item = item;
        if (isEmpty()) {
            first = last = newNode;
            last.prev = null;
            last.next = null;
        } else {
            newNode.prev = last;
            last.next = newNode;
            last = newNode;
            newNode.next = null;
        }
        length++;
    }

    public void insertAtPos1(int item, int pos) {
        DoublyNode newNode = new DoublyNode();
        newNode.item = item;
        if (isEmpty()) {
            first = last = newNode;
            first.prev = null;
            last.next = null;
        } else if (pos == 0) {
            insertToFirst(item);
        } else if (pos == length - 1) {
            insertToLast(item);
        } else {
            DoublyNode cur = first;
            int count = 0;
            for (int i = 0; i != pos - 1; i++) {
                cur = cur.next;
                count++;
            }
            cur.next.prev = newNode;
            newNode.next = cur.next;
            cur.next = newNode;
            newNode.prev = cur;
            length++;
            System.out.println(count);
        }
    }

    public void insertAtPos2(int item, int pos) {
        DoublyNode newNode = new DoublyNode();
        newNode.item = item;
        if (isEmpty()) {
            first = last = newNode;
            first.prev = null;
            last.next = null;
        } else if (pos == 0) {
            insertToFirst(item);
        } else if (pos == length - 1) {
            insertToLast(item);
        } else if (pos <= length / 2) {
            DoublyNode cur = first;
            int count = 0;
            for (int i = 0; i != pos - 1; i++) {
                cur = cur.next;
                count++;
            }
            cur.next.prev = newNode;
            newNode.next = cur.next;
            cur.next = newNode;
            newNode.prev = cur;
            length++;
            System.out.println(count);
        } else if (pos > length / 2) {
            DoublyNode cur = last;
            int count = 0;
            for (int i = 0; i != length - pos; i++) {
                cur = cur.prev;
                count++;
            }
            cur.next.prev = newNode;
            newNode.next = cur.next;
            cur.next = newNode;
            newNode.prev = cur;
            length++;
            System.out.println(count);
        }
    }

    public void removeFromFirst() {
        if (isEmpty()) {
            System.out.println("is already empty");
        } else if (length == 1) {
            first = last = null;
            length--;
        } else {
            first = first.next;
            first.prev = null;
            length--;
        }
    }

    public void removeFromLast() {
        if (isEmpty()) {
            System.out.println("is already empty");
        } else if (length == 1) {
            first = last = null;
            length--;
        } else {
            last = last.prev;
            last.next = null;
            length--;
        }
    }

    public void removeAtElement(int item) {
        if (isEmpty()) {
            System.out.println("is already empty");
        } else if (first.item == item || length == 1) {
            if (first.item == item) {
                removeFromFirst();
            } else {
                System.out.println("the element cant found");
            }
        } else if (last.item == item) {
            removeFromLast();
        } else {
            DoublyNode cur = first.next;
            while (cur != last) {
                if (cur.item == item) {
                    cur.prev.next = cur.next;
                    cur.next.prev = cur.prev;
                    length--;
                    return;
                }
                cur = cur.next;
            }
            System.out.println("the element cant found");
        }
    }

    public int search(int element) {
        DoublyNode cur = first;
        int pos = 0;
        if (isEmpty()) {
            return 0;
        } else {
            while (cur != null) {
                if (cur.item == element) {
                    return pos;
                }
                pos++;
                cur = cur.next;
            }
        }
        return -1;
    }

    public int getMid() {
        DoublyNode curFirst = first;
        DoublyNode curLast = last;
        while (curFirst != curLast) {
            curFirst=curFirst.next;
            curLast=curLast.prev;
        }
        return curFirst.item;
       
    }

    public void printLinkdList() {
        DoublyNode cur = first;
        while (cur != null) {
            System.out.print(cur.item + "  ");
            cur = cur.next;
        }
        System.out.println();
    }

    public void printLinkdListReverse() {
        DoublyNode cur = last;
        while (cur != null) {
            System.out.print(cur.item + "  ");
            cur = cur.prev;
        }
        System.out.println();
    }

    public boolean isEmpty() {
        return length == 0;
    }

    public int getLast() {
        return last.item;
    }

    public int getFirst() {
        return first.item;
    }

    public int getLength() {
        return length;
    }

}

File: DataStructures/c++/doublyLinkedList/doublyLinkedList.cpp
#include <iostream>
#include <cmath>
using namespace std;

template <class T>

class DoublyLinkedList
{
private:
    struct Node
    {
        T item;
        Node *next;
        Node *prev;
    };
    Node *first;
    Node *last;
    int length;

public:
    DoublyLinkedList()
    {
        first = NULL;
        last = NULL;
        length = 0;
    }

    void insertToFirst(T item)
    {
        Node *newNode = new Node;
        newNode->item = item;
        if (isEmpty())
        {
            first = last = newNode;
            first->next = NULL;
            first->prev = NULL;
        }
        else
        {
            newNode->next = first;
            first->prev = newNode;
            first = newNode;
            first->prev = NULL;
        }
        length++;
    }

    void insertToLast(T item)
    {
        Node *newNode = new Node;
        newNode->item = item;
        if (isEmpty())
        {
            last = first = newNode;
            last->next = NULL;
            last->prev = NULL;
        }
        else
        {
            newNode->prev = last;
            last->next = newNode;
            last = newNode;
            last->next = NULL;
        }
        length++;
    }

    void insertAtPos(int pos, T item)
    {
        Node *newNode = new Node;
        Node *cur = first;
        newNode->item = item;
        if (pos < 0 || pos > length)
        {
            cout << "out of range" << endl;
        }
        else if (pos == 0)
        {
            insertToFirst(item);
        }
        else if (pos == length)
        {
            insertToLast(item);
        }
        else
        {
            for (int i = 0; i < pos - 1; i++)
            {
                cur = cur->next;
            }
            newNode->next = cur->next;
            newNode->prev = cur;
            newNode->next->prev = newNode;
            newNode->prev->next = newNode;
            length++;
        }
    }

    void removeFromFirst()
    {
        if (isEmpty())
        {
            cout << "dose not have a items to remove" << endl;
        }
        else if (length == 1)
        {
            first = last = NULL;
            length--;
        }
        else
        {
            Node *cur = first;
            first = first->next;
            first->prev = NULL;
            cur = NULL;
            length--;
        }
    }

    void removeFromLast()
    {
        if (isEmpty())
        {
            cout << "dose not have a items to remove" << endl;
        }
        else if (length == 1)
        {
            first = last = NULL;
            length--;
        }
        else
        {
            Node *cur = last;
            last = last->prev;
            last->next = NULL;
            cur = NULL;
            length--;
        }
    }

    void removeAtElement(T element)
    {
        Node *cur = first;
        if (isEmpty())
        {
            cout << "dose not have a items to remove" << endl;
        }
        else if (first->item == element)
        {
            removeFromFirst();
        }
        else if (last->item == element)
        {
            removeFromLast();
        }
        else
        {
            Node *cur = first;
            while (cur->item != element)
            {
                cur = cur->next;
                if(cur == NULL){
                    cout<<"the element cant found"<<endl;
                    return;
                }
            }
            cur->next->prev = cur->prev;
            cur->prev->next = cur->next;
            cur == NULL;
            length--;
        }
    }

     //! الREVERSE بال doubly عالفاضي اصلا 
    void reverse()
    {

        if (isEmpty())
        {
            cout << "dose not have a items to reverse" << endl;
        }
        else
        {
            Node *prev = NULL;
            Node *cur = first;
            Node *next = first->next;
            while (cur != NULL)
            {
                next = cur->next;
                cur->prev = next;
                cur->next = prev;
                prev = cur;
                cur = next;
            }
            last = first;
            first = prev;
        }
    }

    int search(T element)
    {
        Node *cur = first;
        int pos = 0;
        while (cur != NULL)
        {
            if (cur->item == element)
            {
                return pos;
            }
            cur = cur->next;
            pos++;
        }
        return -1;
    }

    bool isEmpty()
    {
        return length == 0;
    }

    void printStackReverse()
    {
        Node *cur = last;
        cout << "[ ";
        while (cur != NULL)
        {
            cout << cur->item << " ";
            cur = cur->prev;
        }
        cout << "]";
        cout << endl;
    }

    void printStack()
    {
        Node *cur = first;
        cout << "[ ";
        while (cur != NULL)
        {
            cout << cur->item << " ";
            cur = cur->next;
        }
        cout << "]";
        cout << endl;
    }
    int getLength()
    {
        return length;
    }
    void printNULL()
    {
        cout << "first->prev:" << first->prev << "\tlast->next:" << last->next << endl;
    }
};

int main()
{
    DoublyLinkedList<int>d;
    d.insertToFirst(1);
    d.insertToFirst(2);
    d.insertToFirst(3);
    d.insertToFirst(4);
    d.insertToFirst(5);
    d.insertToLast(6);
    d.printStack();
    d.removeAtElement(1);
    d.printStack();
}